File: leetcode/20.py
class Solution(object):
    def isValid(self, s):
        parantheses = {
            "(": ")",
            "[": "]",
            "{": "}"
        }

        answer = []

        for i in s:
            if i in parantheses.keys():
                answer.append(i)
            else:
                if not answer or parantheses[answer.pop()] != i:
                    return False

        if answer:
            return False
        else:
            return True


class Solution(object):
    def isValid(self, s):
        parantheses = {
            ")": "(",
            "]": "[",
            "}": "{"
        }
        answer = []

        for i in s:
            if i not in parantheses:
                answer.append(i)
            elif not answer or parantheses[i] != answer.pop():
                return False

        return len(answer) == 0

File: leetcode/20.js
var isValid = function(s) {
    const stack = []
    const paren = {
        ")" : "(",
        "}" : "{",
        "]" : "["
    }
    const keys = Object.keys(paren);
    
    for (let i=0; i<s.length; i++){
        if(!keys.includes(s[i])){
            stack.push(s[i])
        }else{
            if(!stack || paren[s[i]] !== stack.pop()){
                return false
            }
        }
    }
    return stack.length === 0
};
